Renew gloo certificates before expiration

Megum1n opened the pull request:
# Description

This change will make certgen job renew existing certificate if it is expiring in two months.

Issues: https://github.com/solo-io/gloo/issues/5944 https://github.com/solo-io/gloo/issues/6105

## Code changes
- Update `GetExistingValidTlsSecret()` function to return `nil` in cert is expiring soon
- Add `renew-before` flag to be able to configure certificate rotation
- Add ability to set `force-rotation` flag via helm chart
- Create certgen cronjob if validation webhook is enabled 

# Context

New resources cannot be deployed in k8s cluster if validation webhook does not work. If gloo certificate is expired and gloo validation webhook is deployed to the cluster - new resources cannot be applied.
```plaintext
Upstream/test-ns/test-service dry-run failed, reason: InternalError, error: Internal error occurred: failed calling webhook "gloo.gloo-system.svc": failed to call webhook: Post "https://gloo.gloo-system.svc:443/validation?timeout=10s": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2023-09-04T12:45:55Z is after 2023-08-31T11:40:47Z
```
To avoid this issue, certificate should be renewed before expiration date.

## Testing steps

I manually created certgen job in my cluster and old certificate (Oct 2023 expiration date) was renewed.

## Notes for reviewers

~~We can go further and add configuration option for renewal period. Also, we can create cronjob for gateway certgen which will perform cert validation every X days/months (cron schedule). Current solution requires someone to perform helm upgrade, which executes pre-upgrade hook with certgen job in it.~~
Addressed it in the latest changes.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5944

nfuden commented on jobs/pkg/kube/secret.go:
```diff
@@ -60,7 +60,8 @@ func GetExistingValidTlsSecret(ctx context.Context, kube kubernetes.Interface, s
 
-		if now.Before(cert.NotBefore) || now.After(cert.NotAfter) {
+		// Create new certificate if old one is expiring soon (two months)
+		if now.Before(cert.NotBefore) || now.After(cert.NotAfter.AddDate(0, -2, 0)) {
```
This is pretty opinionated, i think its a reasonable default but it pretty much has to be configurable imo.

nfuden commented:
Thanks for the contribution! May I ask where the 2 months number is coming from?

elcasteel commented on install/helm/gloo/values-template.yaml:
```diff
@@ -90,6 +90,7 @@ gateway:
     setTtlAfterFinished: true
     ttlSecondsAfterFinished: 60
     runAsUser: 10101
+    forceRotation: false
```
nit: make this default to true to match the current behavior

elcasteel reviewed (commented):
Overall looks great, thanks! One tiny comment.

jenshu commented on install/helm/gloo/templates/19-gloo-certgen-cronjob.yaml:
```diff
@@ -60,13 +60,15 @@ spec:
                 - "--secret-name=gloo-mtls-certs"
                 - "--svc-name=gloo"
+                {{- if .Values.gateway.certGenJob.forceRotation }}
                 - "--force-rotation=true"
-{{- end }} {{/* if and .Values.global.glooMtls.enabled .Values.gateway.certGenJob.cron.enabled */}}
+                {{- end }} {{/* if .Values.gateway.certGenJob.forceRotation */}}
```
i think we would want the cronjob to always force rotation, rather than be guarded by a flag

Megum1n replied:
@jenshu I'm not sure about this one

- Force rotation is enabled by default
- There is a new `renew-before` flag now which is set to `2160h` (90 days). Any certificate will be renewed if its expiration time is due soon

I think that more configuration options for users is a better solution than forcing renewal for everyone.

jenshu commented on jobs/pkg/kube/secret.go:
```diff
@@ -60,7 +60,8 @@ func GetExistingValidTlsSecret(ctx context.Context, kube kubernetes.Interface, s
 		}
 
-		if now.Before(cert.NotBefore) || now.After(cert.NotAfter) {
+		// Create new certificate if old one is expiring soon (two months)
```
can remove the "two months" since it's now configurable

jenshu reviewed (dismissed):
just one more comment, but approving. thanks for the PR!

Megum1n commented on install/helm/gloo/templates/19-gloo-certgen-cronjob.yaml:
```diff
@@ -1,9 +1,9 @@
 {{- $image = merge .Values.gateway.certGenJob.image .Values.global.image }}
 {{- end }}
-{{- if and .Values.global.glooMtls.enabled .Values.gateway.certGenJob.cron.enabled }}
+{{- if and (or .Values.global.glooMtls.enabled .Values.gateway.validation.webhook.enabled) .Values.gateway.certGenJob.cron.enabled }}
```
I just found out that this cronjob is for mtls certificates only because of `--secret-name=gloo-mtls-certs` flag.
Certgen is needed to be executed  with `--secret-name={{ .Values.gateway.validation.secretName }}` flag as well. Seems like small refactoring is required :)

Options:
- Add second container in the same cronjob which will be added if `.Values.gateway.validation.enabled`
- Add second separate cronjob for "validation secret"

First option is better imo. What do you think?

jenshu replied:
ahh good catch. there are also some other things like the CronJob name, that are hardcoded with "mtls".  my preference would be to add a completely separate CronJob for the validation webhook certgen. even though there might be some duplication, they are used for different purposes, and it would make the templates more readable

Megum1n replied:
@jenshu is it okay for two cronjobs to use common `.Values.gateway.certGenJob.cron` options or should I separate them?

jenshu replied:
yeah it's unfortunate that the mtls and vwh certgen jobs share the same `certGenJob` helm values, they really should probably be separate. for now it's probably ok to keep them both using the same `cron` settings since if we want to separate them, there's a bigger refactoring we would have to do

jenshu commented on install/helm/gloo/templates/6.5-gateway-certgen-cronjob.yaml:
```diff
@@ -0,0 +1,75 @@
+{{/* Render template with yaml overrides */}}
+{{- $kubeResourceOverride := dict -}}
+{{- if .Values.gateway.certGenJob -}}
+{{- $kubeResourceOverride = .Values.gateway.certGenJob.cron.KubeResourceOverride -}}
```
typo probably from find-and-replace - `KubeResourceOverride`
we should define a new value `kubeResourceOverride` for the validation webhook cronjob

Megum1n replied:
No a typo, I just used same resources as defined in `6.5-gateway-certgen-job.yaml`

Megum1n replied:
Added new variable

jenshu commented on install/helm/gloo/generate/values.go:
```diff
@@ -451,9 +452,10 @@ Scheduling:
+	Enabled                               *bool                  `json:"enabled,omitempty" desc:"enable the cronjob"`
+	Schedule                              *string                `json:"schedule,omitempty" desc:"Cron job scheduling"`
+	MtlsKubeResourceOverride              map[string]interface{} `json:"mtlsKubeResourceOverride,omitempty" desc:"override fields in the gloo-mtls-certgen cronjob."`
+	ValidationWebhookKubeResourceOverride map[string]interface{} `json:"ValidationWebhookKubeResourceOverride,omitempty" desc:"override fields in the gateway-certgen cronjob."`
```
```suggestion
	ValidationWebhookKubeResourceOverride map[string]interface{} `json:"validationWebhookKubeResourceOverride,omitempty" desc:"override fields in the gateway-certgen cronjob."`
```
(should be lowercase)

jenshu commented on install/helm/gloo/templates/6.5-gateway-certgen-cronjob.yaml:
```diff
@@ -0,0 +1,75 @@
+metadata:
+  labels:
+    app: gloo
+    gloo: gloo-certgen-cronjob
+  name: gloo-certgen-cronjob
```
nit: to match the name of the Job `gateway-certgen`, this could be renamed to `gateway-certgen-cronjob`

jenshu commented on install/helm/gloo/templates/6.5-gateway-certgen-cronjob.yaml:
```diff
@@ -0,0 +1,75 @@
+{{/* Render template with yaml overrides */}}
+{{- $kubeResourceOverride := dict -}}
+{{- if .Values.gateway.certGenJob -}}
+{{- $kubeResourceOverride = .Values.gateway.certGenJob.cron.ValidationWebhookKubeResourceOverride -}}
```
```suggestion
{{- $kubeResourceOverride = .Values.gateway.certGenJob.cron.validationWebhookKubeResourceOverride -}}
```

jenshu commented on install/helm/gloo/templates/6.5-gateway-certgen-job.yaml:
```diff
@@ -69,6 +72,6 @@ spec:
 {{- $kubeResourceOverride := dict -}}
 {{- if .Values.gateway.certGenJob -}}
-{{- $kubeResourceOverride = .Values.gateway.certGenJob.kubeResourceOverride -}}
+{{- $kubeResourceOverride = .Values.gateway.certGenJob.ValidationWebhookKubeResourceOverride -}}
```
this change should be reverted
```suggestion
{{- $kubeResourceOverride = .Values.gateway.certGenJob.kubeResourceOverride -}}
```

jenshu reviewed (commented):
the changes look good to me! 2 questions:
- do you think there might be a need for the jobs and cronjobs to have different `forceRotation` values? (e.g. `.Values.gateway.certGenJob.forceRotation` and `.Values.gateway.certGenJob.cron.forceRotation`)
- can you add tests for the new gateway-certgen-cronjob in `install/test/helm_test.go`?  (can probably just insert a new entry in all the places the mtls cronjob is tested)

Megum1n commented:
@jenshu 

- `forceRotation` option will be used not only for cronjobs, but also for helm-hook jobs. I don't think that there is a need for separate configuration for each job. I can create a separate issue to consider configuration split for each job if you want. For now, I suggest we leave it as is.

- Thank you, I added new tests. Also added forceRotation option to the test. 
Sorry, my Goland formatted the whole file a bit, I hope you do not mind.

jenshu commented:
/test

jenshu commented:
> * `forceRotation` option will be used not only for cronjobs, but also for helm-hook jobs. I don't think that there is a need for separate configuration for each job. I can create a separate issue to consider configuration split for each job if you want. For now, I suggest we leave it as is.

I agree we can let them share the same values for now, and if the need arises we can split them later

elcasteel commented:
/kick looks like something in docker

jenshu commented:
@Megum1n sorry about that. to get around this, can you please remove the offending `echo "WARNING"...` from the Makefile?

jenshu commented:
/test

jenshu commented on install/test/helm_test.go:
```diff
@@ -6357,6 +6364,7 @@ metadata:
 					Entry("6-access-logger-deployment", "accessLogger.deployment.kubeResourceOverride", "accessLogger.enabled=true"),
 					Entry("6-access-logger-service", "accessLogger.service.kubeResourceOverride", "accessLogger.enabled=true"),
 					Entry("6.5-gateway-certgen-job", "gateway.certGenJob.kubeResourceOverride"),
+					Entry("6.5-gateway-certgen-cronjob", "gateway.certGenJob.kubeResourceOverride"),
```
```suggestion
					Entry("6.5-gateway-certgen-cronjob", "gateway.certGenJob. cron.validationWebhookKubeResourceOverride"),
```

jenshu replied:
i'm not sure but might also need to add the other values like `gateway.validation.enabled=true`, etc

jenshu commented:
/test

jenshu commented on install/test/helm_test.go:
```diff
@@ -6357,6 +6364,7 @@ metadata:
 					Entry("6-access-logger-deployment", "accessLogger.deployment.kubeResourceOverride", "accessLogger.enabled=true"),
 					Entry("6-access-logger-service", "accessLogger.service.kubeResourceOverride", "accessLogger.enabled=true"),
 					Entry("6.5-gateway-certgen-job", "gateway.certGenJob.kubeResourceOverride"),
+					Entry("6.5-gateway-certgen-cronjob", "gateway.certGenJob.cron.validationWebhookKubeResourceOverride"),
```
probably need to add the other values `"gateway.enabled=true", "gateway.validation.enabled=true", "gateway.validation.webhook.enabled=true", "gateway.certGenJob.cron.enabled=true"`

elcasteel commented:
/test
